homeassistant/components/mjpeg/config_flow.py

```diff
@@ -0,0 +1,250 @@
+
+from homeassistant.config_entries import ConfigEntry, ConfigFlow, OptionsFlow
+from homeassistant.const import (
+    CONF_AUTHENTICATION,
+    CONF_NAME,
+    CONF_PASSWORD,
+    CONF_USERNAME,
+    CONF_VERIFY_SSL,
+    HTTP_BASIC_AUTHENTICATION,
+    HTTP_DIGEST_AUTHENTICATION,
+)
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.exceptions import HomeAssistantError
+
+from .const import CONF_MJPEG_URL, CONF_STILL_IMAGE_URL, DOMAIN, LOGGER
+
+
+@callback
+def async_get_schema(
+    defaults: dict[str, Any] | MappingProxyType[str, Any], show_name: bool = False
+) -> vol.Schema:
+    """Return MJPEG IP Camera schema."""
+    schema = {
+        vol.Required(CONF_MJPEG_URL, default=defaults.get(CONF_MJPEG_URL)): str,
+        vol.Optional(
+            CONF_STILL_IMAGE_URL,
+            description={"suggested_value": defaults.get(CONF_STILL_IMAGE_URL)},
+        ): str,
+        vol.Optional(
+            CONF_USERNAME,
+            description={"suggested_value": defaults.get(CONF_USERNAME)},
+        ): str,
+        vol.Optional(
+            CONF_PASSWORD,
+            default=defaults.get(CONF_PASSWORD, ""),
+        ): str,
+        vol.Optional(
+            CONF_VERIFY_SSL,
+            default=defaults.get(CONF_VERIFY_SSL, True),
+        ): bool,
+    }
+
+    if show_name:
+        schema = {
+            vol.Optional(CONF_NAME, default=defaults.get(CONF_NAME)): str,
+            **schema,
+        }
+
+    return vol.Schema(schema)
+
+
+def validate_url(
+    url: str,
+    username: str | None,
+    password: str,
+    verify_ssl: bool,
+    authentication: str = HTTP_BASIC_AUTHENTICATION,
+) -> str:
+    """Test if the given setting works as expected."""
+    auth: HTTPDigestAuth | HTTPBasicAuth | None = None
+    if username and password:
+        if authentication == HTTP_DIGEST_AUTHENTICATION:
+            auth = HTTPDigestAuth(username, password)
+        else:
+            auth = HTTPBasicAuth(username, password)
+
+    response = requests.get(
+        url,
+        auth=auth,
+        stream=True,
+        timeout=10,
+        verify=verify_ssl,
+    )
+
+    if response.status_code == HTTPStatus.UNAUTHORIZED:
+        # If unauthorized, try again using digest auth
+        if authentication == HTTP_BASIC_AUTHENTICATION:
+            return validate_url(
+                url, username, password, verify_ssl, HTTP_DIGEST_AUTHENTICATION
+            )
+        raise InvalidAuth
+
+    response.raise_for_status()
+    response.close()
+
+    return authentication
+
+
+async def async_validate_input(
+    hass: HomeAssistant, user_input: dict[str, Any]
+) -> tuple[dict[str, str], str]:
+    """Manage MJPEG IP Camera options."""
+    errors = {}
+    field = "base"
+    authentication = HTTP_BASIC_AUTHENTICATION
+    try:
+        # Validate MJPEG URL
+        field = CONF_MJPEG_URL
+        authentication = await hass.async_add_executor_job(
```

Comment:
This could be a bit more DRY if you make both urls loop vars

Reply:
Little more DRY: [7fb5161](https://github.com/home-assistant/core/pull/66607/commits/7fb51616f87f937b9851f713af22ddd914608e7b)